### PR TITLE
[Reland][ONNX] Guard xfail tests with error messages

### PR DESCRIPTION
**Pull request by titaiwangms**
Reland #117425 

Previous to this PR, xfail tests didn't provide (1) guarantee of error message/reason (could be outdated), and (2) execution of the test (xfail_if_model_type_is_not_exportedprogram). Therefore, the tests are less robust with xfail labeled, as we can't be sure if it's still failing with the same reason, and if it's even still failing. This PR fixes the issue with try/except with error message matching to consolidate the xfail truth and reason.

### CODE REVIEW

**Review by BowenBao — APPROVED** [review]

**Review by thiagocrepaldi — COMMENTED** [review]
Thanks

File: test/onnx/pytorch_test_common.py [hunk 1]
```diff
@@ -292,7 +274,7 @@ def wrapper(self, *args, **kwargs):
     return skip_dec
 
 
-def xfail(reason: str):
+def xfail(error_message: str, reason: Optional[str] = None):
```
Comment (justinchuby) [inline test/onnx/pytorch_test_common.py:277]: document the arg?

File: test/onnx/pytorch_test_common.py [hunk 2]
```diff
@@ -301,7 +283,29 @@ def xfail(reason: str):
...
+        @functools.wraps(func)
+        def inner(self, *args, **kwargs):
+            try:
+                func(self, *args, **kwargs)
+            except Exception as e:
+                if isinstance(e, torch.onnx.OnnxExporterError):
+                    # diagnostic message is in the cause of the exception
+                    assert error_message in str(
+                        e.__cause__
+                    ), f"Expected error message: {error_message} NOT in {str(e.__cause__)}"
```
Comment (justinchuby) [inline test/onnx/pytorch_test_common.py:297]: Consider using regex for this? Example api: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex

**Pull request closed without merging**